torch/csrc/DataLoader.cpp

```diff
@@ -0,0 +1,244 @@
+// events happen.
+
+// TODO: The following don't work on Windows. Specifically, waitid calls and
+// SIGCHLD handler. Currently, dummy implementation is provided for Windows.
+
+#ifndef _WIN32
+
+// Critical signal handlers should be registered on worker processes before
+// doing work.
+// Python handle is _set_worker_signal_handlers().
+#define SIGNAL_HANDLER(SIGNAL, HANDLER_NAME, ERROR_MSG)                       \
+static void HANDLER_NAME(int sig, siginfo_t *info, void *ctx)                 \
+{                                                                             \
+    write(STDERR_FILENO, ERROR_MSG, sizeof(ERROR_MSG) / sizeof(char));        \
+    _exit(EXIT_FAILURE);                                                      \
+}
+
+// signal(2) is really not portable. So use sigaction.
+// http://man7.org/linux/man-pages/man2/signal.2.html
+static void setSignalHandler(int signal, void(*handler)(int, siginfo_t *, void *), struct sigaction *old_sa_ptr)
+{
+  struct sigaction sa;
+  sa.sa_sigaction = handler;
+  sa.sa_flags = SA_RESTART|SA_SIGINFO|SA_NOCLDSTOP;
+  sigemptyset(&sa.sa_mask);
+  if (sigaction(signal, &sa, old_sa_ptr) != 0) {
+    std::ostringstream oss;
+    oss << "An error occurred while setting handler for " << strsignal(signal);
+    throw std::runtime_error(oss.str());
+  }
+}
+
+SIGNAL_HANDLER(SIGBUS, handler_SIGBUS, "ERROR: Unexpected bus error encountered in worker. "
+  "This might be caused by insufficient shared memory (shm).\n");
+SIGNAL_HANDLER(SIGSEGV, handler_SIGSEGV, "ERROR: Unexpected segmentation fault encountered in worker.\n");
+
+PyObject *THPModule_setWorkerSignalHandlers(PyObject *module, PyObject *arg) {
+  HANDLE_TH_ERRORS
+  setSignalHandler(SIGBUS, &handler_SIGBUS, NULL);
+  setSignalHandler(SIGSEGV, &handler_SIGSEGV, NULL);
+  Py_RETURN_TRUE;
+  END_HANDLE_TH_ERRORS
+}
+
+static std::set<pid_t> worker_pid_set = {};
+// The following are needed since std::set is not asynchronous safe.
+static std::atomic<pid_t *> worker_pids;
+static std::atomic<size_t> num_worker_pids(0);
+// Pipe used as a lock to avoid update of the above and SIGCHLD handler in parallel.
+static int comm_pipe[2] = {-1, -1};
+
+static void updatePIDsArray() {
+  size_t new_size =  worker_pid_set.size();
+  auto new_ptr = (pid_t *)malloc(sizeof(pid_t) * new_size);
+  size_t idx = 0;
+  for (auto it = worker_pid_set.begin(); it != worker_pid_set.end(); it++, idx++) {
+    new_ptr[idx] = *it;
+  }
+
+  // Block SIGCHLD handler for this thread so SIGCHLD handler can't interrupt
+  // from this thread
+  sigset_t sigset, old_sigset;
+  sigemptyset(&sigset);
+  sigaddset(&sigset, SIGCHLD);
+  if (sigprocmask(SIG_BLOCK, &sigset, &old_sigset) != 0) {
+    throw std::runtime_error("An error occurred while setting worker information "
+      "for DataLoader SIGCHLD handler");
+  }
+  // Acquire ``lock'' so handlers on other threads can't interrupt
+  char c;
+  read(comm_pipe[0], &c, 1);
```

Comment:
It would be nice to encapsulate this pipe in a C++ struct that conforms to the mutex protocol, so we could actually use `lock_guard`s instead of those calls to `read` and `write`

Reply:
BTW I'm pretty sure this this locking is potentially susceptible to deadlocks - if your process is limited to a single core, the signal handler might interrupt this function while it holds the "lock", but it won't be able to acquire it, and will wait forever. In general, I think the only sane solution we could use is `std::atomic` for the pointer and size.

Reply:
Good point. I'm already using `std:atomic`, but I think I still need to think of how to properly free old array pointer.